### PR TITLE
Fix link color for roles without unfiltered_html capabilities

### LINKED ISSUE #25151: Link color doesn't work for author role
How to test:

- Install the [Seedlet theme](https://wordpress.org/themes/seedlet/), or any other theme that has support for the [link color feature](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#experimental-%e2%80%94-link-color-control).
- As an author role, go to the post editor and add a paragraph with a link.
- Change the link color via the block settings.
- Publish post.

The expected result is that the paragraph, in the front-end, has the class `has-link-color` and also has inlined the style for the color such as `style="--wp--style--color--link:<some-color-here>"`. What happens instead is that the CSS custom property isn't present.

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/25151

### CODE REVIEW

File: lib/global-styles.php
Comment (nosolosw) [inline lib/global-styles.php:883]: kses has several checks when filtering post content, [this](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2150) is the relevant function when it comes to filter the `style` attribute:

- check whether the property name is one of the allowed ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2177))
- check that the property value doesn't have extraneous characters, excluding parentheses as well ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2381))

File: lib/global-styles.php
Comment (nosolosw) [inline lib/global-styles.php:877]: mmm, I believe we want to improve this bit. Allowing any characters (`.`) may still be problematic security-wise. It feels like we want to allow only alphanumeric and `-` as valid characters (eg of potential names: `...--some-color-value`, `...--some-color-value-2`, etc), which should cover all the potential color names that people use.

Comment (nosolosw) [reply lib/global-styles.php]: improved at https://github.com/WordPress/gutenberg/pull/25411/commits/d27c8698c322aa007664861ac2ef37878104907b

File: lib/global-styles.php
Comment (jorgefilipecosta) [inline lib/global-styles.php:870]: Should we apply trim before comparing to remove whitespace?

File: lib/global-styles.php
Comment (jorgefilipecosta) [inline lib/global-styles.php:877]: I guess trim here may also be helpful.

Comment (jorgefilipecosta) [inline lib/global-styles.php:877]: We are only accepting the variable values for the link color, which means that currently, custom link colors are not working.

I guess the custom color values are already accepted so we can fix the issue by simply return true if $allow_css is already true. Meaning we will not apply any additional restriction if an attribute is accepted we will continue to accept it but we will also accept the CSS variable references.

**Review by jorgefilipecosta — COMMENTED** [review]
In the core, we already have a good testing infrastructure for kses rule changes. And we will need to port this change to core anyways. Would it make sense the create a core patch with the change here and adding some kses rule test cases there? Similar to the patch we previously implemented for gradients https://github.com/WordPress/wordpress-develop/pull/110/files.

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php]: https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

File: lib/global-styles.php
Comment (nosolosw) [reply lib/global-styles.php]: https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

Comment (nosolosw) [reply lib/global-styles.php]: Good catch, thanks!

Comment (nosolosw) [reply lib/global-styles.php]: fixed at https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

**Comment by nosolosw** [comment]
> In the core, we already have a good testing infrastructure for kses rule changes. And we will need to port this change to core anyways. Would it make sense the create a core patch with the change here and adding some kses rule test cases there?

I can do that as a follow-up to this. I'd also want to merge this as it is to make sure the link color is no longer broken in the plugin without having to wait for a core release that is still a few months away.

What I'm unsure about is: _when_ to prepare/merge that core patch. Is the 5.6 window a good time? I know there's uncertainty about whether link color is going to be merged into core in 5.6 (it doesn't look like it) so I wonder if we should wait to the 5.7 window instead. To my understanding, people have talked about that a few things can change (the general approach -classes instead of css vars-, the name of the var, the name of the value). I haven't done a core patch before so I don't have a good sense of whether this is a good idea or not (things like how difficult would it be to remove/update the patch, should we need it). So, essentially, I'm happy to take any advice here. What do you think?

**Review by jorgefilipecosta — APPROVED** [review]
LGTM 👍
The reason I referred the core patch was to be able to test this kses changes against the core kses tests to increase the confidence we have on this change given that it may have a security impact. It would be good to test the patch before the next Gutenberg release.

**Comment by nosolosw** [comment]
I'm trying to figure out who can provide a sanity check from a security point of view before this is merged.

**Comment by nosolosw** [comment]
I've run this by some security folks and they aren't wild about serializing CSS Custom Properties in the post content. They'd rather keep the behavior as it is (not allowing CSS vars in the post content) or allow any in core, as it was done for `data-*` attributes in the past.

So we're back at the drawing board. Going to close this for the moment. Perhaps we can revisit Riad's https://github.com/WordPress/gutenberg/pull/21420

**Pull request closed without merging**
